File: Diagrammatize.Design.Pattern/code/cxx/Interpreter/A1/language/PrimitiveCommandNode.cpp
#include "PrimitiveCommandNode.hpp"
#include "ParseException.hpp"
#include "ExecuteException.hpp"

PrimitiveCommandNode::~PrimitiveCommandNode() {}

void PrimitiveCommandNode::parse(Context& context) {
    name_ = context.currentToken();
    context.skipToken(name_);
    executor_ = context.createExecutor(name_);
}

std::string PrimitiveCommandNode::toString() {
    return name_;
}

void PrimitiveCommandNode::execute() {
    if (!executor_) {
        throw ExecuteException(name_ + ": is not defined");
    } else {
        executor_->execute();
    }
}

File: Diagrammatize.Design.Pattern/code/python/Interpreter/A1/language/PrimitiveCommandNode.py

from .Node import Node
from .ParseException import ParseException

# <primitive command> ::= go | right | left
class PrimitiveCommandNode(Node):
    def __init__(self):
        super().__init__()
        self.name = None
        self.executor = None
    
    def parse(self, context):
        self.name = context.currentToken()
        context.skipToken(self.name)
        self.executor = context.createExecutor(self.name)
    
    def execute(self):
        if self.executor is None:
            raise ExecuteException(f"{self.name}: is not defined")
        else:
            self.executor.execute()
    
    def __str__(self):
        return self.name
